provider/source/src/units/data.rs

```diff
@@ -32,70 +32,52 @@ impl DataProvider<UnitsDisplayNameV1Marker> for SourceDataProvider {
-            ) {
-                (Some(length), None) => length,
-                (None, Some(length)) => length,
-                _ => {
-                    return Err(DataErrorKind::IdentifierNotFound
-                        .into_error()
-                        .with_debug_context(unit))
-                }
-            };
-
-            add_unit_to_map_with_name(&mut result, Count::One, unit_length_map.one.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Two, unit_length_map.two.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Few, unit_length_map.few.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Many, unit_length_map.many.as_deref());
-            add_unit_to_map_with_name(&mut result, Count::Other, unit_length_map.other.as_deref());
-
-            Ok(result)
+        .iter()
+        // get the units which match the key after the `-` in the attribute
+        // For exmple,
+        //          if the unit is meter, the key `length-meter` will match.
+        //          if the unit is millisecond, the key `duration-millisecond` will match.
+        .find_map(|(key, patterns)| {
+            key.split_once('-')
+                .map(|(_, rest)| rest)
+                .filter(|&rest| rest == unit)
+                .map(|_| patterns)
+        })
+        .ok_or_else(|| {
+            DataErrorKind::InvalidRequest
+                .into_error()
+                .with_debug_context(length)
+        })?;
+
+        for (count, unit) in [
+            (Count::One, unit_patterns.one.as_deref()),
+            (Count::Two, unit_patterns.two.as_deref()),
+            (Count::Few, unit_patterns.few.as_deref()),
+            (Count::Many, unit_patterns.many.as_deref()),
+            (Count::Other, unit_patterns.other.as_deref()),
+        ] {
+            if let Some(unit) = unit {
+                patterns.insert(count, unit);
+            }
         }
 
-        let result = UnitsDisplayNameV1 {
-            patterns: ZeroMap::from_iter(
-                populate_units_map(
-                    match length {
-                        "long" => &units_format_data.long,
-                        "short" => &units_format_data.short,
-                        "narrow" => &units_format_data.narrow,
-                        _ => {
-                            return Err(DataErrorKind::IdentifierNotFound
-                                .into_error()
-                                .with_debug_context(length))
-                        }
-                    },
-                    unit,
-                )?
-                .iter()
-                .map(|(k, v)| (k, v.as_str())),
-            ),
-        };
-
         Ok(DataResponse {
             metadata: Default::default(),
-            payload: DataPayload::from_owned(result),
+            payload: DataPayload::from_owned(UnitsDisplayNameV1 {
+                patterns: ZeroMap::from_iter(patterns.iter().map(|(k, v)| (k, *v))),
```

Comment:
no need for the intermediate `BTreeMap` really

```suggestion
                patterns: ZeroMap::from_iter([
                    (Count::One, unit_patterns.one.as_deref()),
                    (Count::Two, unit_patterns.two.as_deref()),
                    (Count::Few, unit_patterns.few.as_deref()),
                    (Count::Many, unit_patterns.many.as_deref()),
                    (Count::Other, unit_patterns.other.as_deref()),
                  ].into_iter().filter_map(|(count, unit)| Some((count, unit?)))
                ),
```